roadmap/implementers-guide/src/runtime/router.md

```diff
@@ -144,22 +148,34 @@ HrmpCloseChannelRequests: map HrmpChannelId => Option<()>;
 HrmpCloseChannelRequestsList: Vec<HrmpChannelId>;
 
 /// The HRMP watermark associated with each para.
+/// Invariant: every para should be onboarded.
 HrmpWatermarks: map ParaId => Option<BlockNumber>;
 /// HRMP channel data associated with each para.
 HrmpChannels: map HrmpChannelId => Option<HrmpChannel>;
-/// The indexes that map all senders to their recievers and vise versa.
+/// Ingress/egress indexes allow to find all the senders and receivers given the opposite
+/// side. I.e.
+///
+/// (a) ingress index allows to find all the senders for a given recipient.
+/// (b) egress index allows to find all the recipients for a given sender.
+///
 /// Invariants:
-/// - for each ingress index entry for `P` each item `I` in the index should present in `HrmpChannels` as `(I, P)`.
-/// - for each egress index entry for `P` each item `E` in the index should present in `HrmpChannels` as `(P, E)`.
+/// - for each ingress index entry for `P` each item `I` in the index should present in `HrmpChannels`
+///   as `(I, P)`.
+/// - for each egress index entry for `P` each item `E` in the index should present in `HrmpChannels`
+///   as `(P, E)`.
 /// - there should be no other dangling channels in `HrmpChannels`.
+/// - the vectors are sorted.
 HrmpIngressChannelsIndex: map ParaId => Vec<ParaId>;
 HrmpEgressChannelsIndex: map ParaId => Vec<ParaId>;
 /// Storage for the messages for each channel.
 /// Invariant: cannot be non-empty if the corresponding channel in `HrmpChannels` is `None`.
 HrmpChannelContents: map HrmpChannelId => Vec<InboundHrmpMessage>;
 /// Maintains a mapping that can be used to answer the question:
 /// What paras sent a message at the given block number for a given reciever.
-/// Invariant: The para ids vector is never empty.
+/// Invariants:
+/// - The para ids vector is never empty.
+/// - The para ids vector doesn't contain duplicates.
+/// - The outer vector is sorted ascending by block number and there are no duplicates.
```

Comment:
Ambiguous: for a given block number, would it be legal for the outer vector to contain both `(n, vec1)` and `(n, vec2)` where `vec1 != vec2`? That's the property we get from a naive deduplication. If there cannot be two digests with the same block number, regardless of the associated vectors, then that's a stricter property, and we should mention it specifically.